### PR TITLE
feat: DKG AE patten refactory

### PR DESCRIPTION
**Pull request by maqi**

### CODE REVIEW

File: src/key_gen/mod.rs [hunk 1]
```diff
@@ -415,26 +439,50 @@ impl KeyGen {
...
+        match msg.clone() {
             Message::Initialization {
                 key_gen_id,
                 m,
                 n,
                 member_list,
-            } => self.handle_initialization(rng, m, n, key_gen_id, member_list),
-            Message::Proposal { key_gen_id, part } => self.handle_proposal(key_gen_id, part),
+            } => {
+                let _ = self.caching_messages.insert(msg.id(), msg);
```
Comment (joshuef) [inline src/key_gen/mod.rs]: Is this not us caching an incoming message here? 

I thought `caching_messages` was to be used for messages _we_ created?

File: src/key_gen/mod.rs [hunk 2]
```diff
@@ -727,10 +764,8 @@ impl KeyGen {
         invalid_msg: Vec<u8>,
     ) -> Result<Vec<Message>, Error> {
         if self.phase != Phase::Complaining {
-            return Err(Error::UnexpectedPhase {
-                expected: Phase::Complaining,
-                actual: self.phase,
-            });
+            trace!("To avoid triggering AE pattern, skip this so far");
+            return Ok(Vec::new());
```
Comment (joshuef) [inline src/key_gen/mod.rs:768]: I'm not sure why we'd want to skip this? (My knowledge of DKG is...low... forgive me!)

File: src/key_gen/mod.rs [hunk 3]
```diff
@@ -1048,7 +1084,7 @@ impl KeyGen {
             initalization_accumulator: InitializationAccumulator::new(),
             complaints_accumulator: ComplaintsAccumulator::new(names, threshold),
             pending_complain_messages: Vec::new(),
-            pending_messages: Vec::new(),
+            caching_messages: BTreeMap::new(),
```
Comment (joshuef) [inline src/key_gen/mod.rs]:
```suggestion
            cached_sent_messages: BTreeMap::new(),
```
The more I'm reading through the more i think this needs a rename for clarity

File: src/key_gen/tests.rs [hunk 4]
```diff
@@ -108,6 +126,7 @@ fn all_nodes_being_responsive() -> Result<()> {
 }
 
 #[test]
+// #[ignore = "The DKG AE pattern doesn't cover Phase span including Complain, which this test is required."]
```
Comment (joshuef) [inline src/key_gen/tests.rs]: remove this comment

File: src/key_gen/mod.rs (hunk 1 above)
Comment (lionel-faber) [reply src/key_gen/mod.rs]: We cache messages that we created and messages that come from other participants as well. Both types flow through this function.

File: src/key_gen/mod.rs (hunk 2 above)
Comment (maqi) [reply src/key_gen/mod.rs:768]: To make the `Complaint` message handable, we may have to cache All the previous Initial/Proposal/Acknowledgment messages, which will complex the DKG AE pattern a lot.
As, I mentioned before, the current DKG AE is based on a `simplified DKG phases` , which doesn't force the Complaint to be supported.

File: src/key_gen/mod.rs (hunk 3 above)
Comment (maqi) [reply src/key_gen/mod.rs]: I am not good at naming stuff.
Any suggestion?

Comment (lionel-faber) [reply src/key_gen/mod.rs]: I think `message_cache` sounds reasonable.

**Review by lionel-faber — APPROVED** [review]

**Pull request merged**